### PR TITLE
An attempt to ease the expected transition to Mongo

### PR DESCRIPTION
**Pull request by vaz**
This is my quick-ish attempt to help avoid quite so much of a painful experience for this cohort swapping in Mongo to Tweetr.

All changes are in the `server/` directory.

I did some renaming (for clarity and/or consistency with things seen elsewhere in the program). I changed how the modules were factored, making more use of "modules that export a factory" because that's the only way to avoid a ton of complication here (or getting into stuff like promises, buffering async requests, or other too-advanced things).

* `lib/db.js` was split into two files:
  - `lib/in-memory-db.js` is as simple as it can get. It requires the initial data JSON and exports it. But separating it makes it a lot easier to swap out for, say, a connected Mongo db instance.
  - `lib/tweet-data.js`, which exports an object I'm calling `TweetData`: the interface to the database (same methods, `saveTweet` and `getTweets`
    - the methods in here are already async: Instead of simulating async delay in the express route handler for `GET /tweets/`, it gets simulated inside these methods, which gives them a good reason to have a method signature that includes `callback` (instead of "no apparent reason", like the earlier connect method).
    - I really believe it will be much easier to understand a module where you "pass in a (connected) db and synchronously returning a data-layer object" than it was to understand a module that "exports a connect function that doesn't do anything but has closure over module privates"

* `api/tweets.js` -> `routes/tweets.js`

* `lib/user-helper.js` -> `lib/util/user-helper.js`
* `lib/tweets.json` -> `resources/initial-tweets.json`
* Added `lib/util/simulate-async.js` (just factoring out the setTimeout with a bit of randomness)

All of the coordination magic happens in `index.js`. Everything that needs to be understood (other than actually working with the database, whatever that happens to be) is happening here (and in updating `lib/tweet-data.js` to use the Mongo db.

The intention, and I hope it's clear, is that you just `MongoClient.connect` in `index.js` and then when you have the db, continue on by using it to create the `TweetData` interface, and using that to define the routes. Yes, the routes won't be defined immediately. I don't think that's bad. It's a detail that's as unimportant (relatively speaking) as properly handling shutdown, which we absolutely don't show them how to do right.

Anyway.. Please comment.

For testing, I branched off this one and stole the `public/` directory from Lydia's project (because I had it handy) so you can use that to see it working: <https://github.com/vaz/tweetr/tree/de-confusify-thanks-lydia>

### CODE REVIEW

**Comment by vaz** [comment]
cc: @kvirani, also @interlock since you're lecturing Thursday, not sure if it would impact.

They'd need to merge this to their repos. Or even just be instructed to replace files. It could be an issue if they touched things outside of public but they shouldn't have done so.

I haven't made any curriculum changes but I know there are parts on Thursday that refer to the file structure that would have to change. That'll have to be reviewed first but I'd like comments on if this seems like the right way to go for now first. If I had to I could rename some things back but I think it would be better how I have them here if it's possible.

anyway I'd appreciate another set of eyes with some consideration of Thursdays' activities if anyone is able.

**Review by interlock — APPROVED** [review]
LGTM

**Comment by vaz** [comment]
@interlock: did some renames after chatting with Monica, I don't have write access here though, can we merge? I'd like to get people sorted out while David has their attention already.

**Pull request merged**
